library/core/src/num/saturating.rs

```diff
@@ -0,0 +1,777 @@
+            ///
+            /// # Examples
+            ///
+            /// Basic usage:
+            ///
+            /// ```
+            /// #![feature(saturating_int_impl)]
+            /// use std::num::Saturating;
+            ///
+            #[doc = concat!("assert_eq!(Saturating(10", stringify!($t), ").signum(), Saturating(1));")]
+            #[doc = concat!("assert_eq!(Saturating(0", stringify!($t), ").signum(), Saturating(0));")]
+            #[doc = concat!("assert_eq!(Saturating(-10", stringify!($t), ").signum(), Saturating(-1));")]
+            /// ```
+            #[inline]
+            #[unstable(feature = "saturating_int_impl", issue = "87920")]
+            pub fn signum(self) -> Saturating<$t> {
+                Saturating(self.0.signum())
+            }
+
+            /// Returns `true` if `self` is positive and `false` if the number is zero or
+            /// negative.
+            ///
+            /// # Examples
+            ///
+            /// Basic usage:
+            ///
+            /// ```
+            /// #![feature(saturating_int_impl)]
+            /// use std::num::Saturating;
+            ///
+            #[doc = concat!("assert!(Saturating(10", stringify!($t), ").is_positive());")]
+            #[doc = concat!("assert!(!Saturating(-10", stringify!($t), ").is_positive());")]
+            /// ```
+            #[inline]
+            #[unstable(feature = "saturating_int_impl", issue = "87920")]
+            pub const fn is_positive(self) -> bool {
+                self.0.is_positive()
+            }
+
+            /// Returns `true` if `self` is negative and `false` if the number is zero or
+            /// positive.
+            ///
+            /// # Examples
+            ///
+            /// Basic usage:
+            ///
+            /// ```
+            /// #![feature(saturating_int_impl)]
+            /// use std::num::Saturating;
+            ///
+            #[doc = concat!("assert!(Saturating(-10", stringify!($t), ").is_negative());")]
+            #[doc = concat!("assert!(!Saturating(10", stringify!($t), ").is_negative());")]
+            /// ```
+            #[inline]
+            #[unstable(feature = "saturating_int_impl", issue = "87920")]
+            pub const fn is_negative(self) -> bool {
+                self.0.is_negative()
+            }
+        }
+    )*)
+}
+
+saturating_int_impl_signed! { isize i8 i16 i32 i64 i128 }
+
+macro_rules! saturating_int_impl_unsigned {
+    ($($t:ty)*) => ($(
+        impl Saturating<$t> {
+            /// Returns the number of leading zeros in the binary representation of `self`.
+            ///
+            /// # Examples
+            ///
+            /// Basic usage:
+            ///
+            /// ```
+            /// #![feature(saturating_int_impl)]
+            /// use std::num::Saturating;
+            ///
+            #[doc = concat!("let n = Saturating(", stringify!($t), "::MAX) / Saturating(4", stringify!($t), ");")]
```

Comment:
why use `x / 4` instead of `x >> 2` like everywhere else.

```suggestion
            #[doc = concat!("let n = Saturating(", stringify!($t), "::MAX) >> 2;")]
```

Reply:
I dropped the shift impls because there are no `*::saturating_shl/r` functions like there are `*::wraping_shl/r`.

Reply:
(well there's no `saturating_bitand` either 🤷)

i guess this should be an "Unresolved Questions" then. what should be the behavior of `Saturating(1u32) << 33`?

1. keep it unimplemented (but hard to argue when `rotate_left` is implemented).
2. panic with "attempt to shift left with overflow" like `1u32 << 33`.
3. "saturate" into `Saturating(1u32 << 31)`.
4. "wrap" into `Saturating(1u32 << 1)`.
5. use the infinite-bit interpretation and return 0.

i prefer doing "2" for now which is forward-compatible with 3, 4 or 5.

Reply:
let's change this line to this for now, which doesn't rely on `impl Shr for Saturating`.

```suggestion
            #[doc = concat!("let n = Saturating(", stringify!($t), "::MAX >> 2);")]
```